File: chip8/Stack.go
package chip8

type Stack []uint16

func (stack *Stack) push(value uint16) {
	*stack = append(*stack, value)
}

func (stack *Stack) pop() uint16 {
	if len(*stack) == 0 {
		return 0
	}
	last := (*stack)[len(*stack)-1]
	*stack = (*stack)[:len(*stack)-1]
	return last
}

File: ports/arduino/src/emulator/stack.cpp
#include <stdint.h>
#include "emulator/stack.h"

bool Stack::isEmpty() const {
    return pointer == 0;
}

void Stack::push(uint16_t value) {
    if (pointer+1 < stackSize) {
        data[pointer] = value;
        pointer++;
    }
}

uint16_t Stack::pop() {
    if (pointer > 0) {
        pointer--;
        return data[pointer];
    }

    return 0;
}
